src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +500,358 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
   }
 }
 
+#ifdef FIU_ENABLE
+
+class HttpInstallationFailed : public HttpFake {
+ public:
+  HttpInstallationFailed(const boost::filesystem::path& test_dir_in) : HttpFake(test_dir_in, "hasupdates") {}
+
+  HttpResponse handle_event(const std::string& url, const Json::Value& data) override {
+    (void)url;
+
+    for (auto& event_data : data) {
+      auto event_id = event_data["eventType"]["id"].asString();
+      auto ecu = event_data["event"]["ecu"].asString();
+      auto correlationID = event_data["event"]["correlationId"].asString();
+
+      if (event_id == "EcuInstallationCompleted") {
+        install_completion_status_[ecu] = event_data["event"]["success"].asBool();
+        if (!install_completion_status_[ecu]) {
+          setFlavor("noupdates");
```

Comment:
I still think changing the flavor in the middle of a test is not a good idea. It won't really test what you expect it to, since that will result in unexpected server behavior. Is it still necessary to do something like this here?

Reply:
Yes, I need to prevent a client from trying to install new updates before running aktualizr again (another UptaneCycle) after installation finalization failure. 
It's possible to remove setFlavor("noupdates") and call just aktualizr.Initialize() after the installation finalization failure, but IMHO, this test won't fully corresponds to what would happen in real life scenario, in the end, the server is not supposed to advertise updates installation of which has failed.

Reply:
> IMHO, this test won't fully corresponds to what would happen in real life scenario, in the end, the server is not supposed to advertise updates installation of which has failed.

Ironically, that is exactly what the server currently does. It continually tells us to keep retrying to install the same file that failed to install until someone explicitly cancels the update. But in theory, I agree, that really shouldn't be the case.

But the bigger issue here is that I don't think changing the flavor mid-test will do what you want it to do. I don't think the metadata in question was generated such that that will work, although I could be wrong. I think what would happen is that the http server will start serving metadata that is entirely unrelated to what libaktualizr previously received, and libaktualizr will reject all of it because it doesn't make any sense. Since the keys, hashes, lengths, signatures, and versions will all change, libaktualizr should just reject it all and assume the server got hacked. I don't think that's what you are trying to test here.

Reply:
You're right saying that aktualizr will reject the metadata and doesn't try to do any updates so I can check installation results of the previous installation, otherwise they will be rewritten by results of the new update installation. Anyway, I can get rid of  setFlavor("noupdates") by not running aktualizr.UptaneCycle() for the second time just after the installation finalization failure, it suffices to run just aktualizr.Initialize() in order to verify what's required by this specific test.